Replace Webpacker with importmapped Hotwire as default JavaScript setup

dhh opened the pull request:
A default Rails skeleton [should not have to require the full JavaScript toolchain](https://world.hey.com/dhh/modern-web-apps-without-javascript-bundling-or-transpiling-a20f2755) with Webpack by default. Let's start with an importmapped Hotwire, then make it easy to go from there to Webpack, if needed later. Though we'll retain --webpack as an option for everyone who knows they'll need the full JS chain (like if doing React).

The new options are as follows:

1. `rails new` produces a skeleton configured with importmap-rails, turbo-rails, and stimulus-rails. No Webpack, no Yarn, no package.json, no node. All the integration setup is done through adding the gems, then running `importmap:install`, `turbo:install`, `stimulus:install` tasks. 
2. `rails new —skip-hotwire` a skeleton like above but with just importmap-rails configured.
3. `rails new —skip-javascript` no JavaScript setup of any kind is configured.
4. `rails new —webpack` is essentially what we had on Rails 6.2, with the exception that the default Turbolinks configuration is replaced with Hotwire (both Turbo and Stimulus). This of course includes setting up package.json, Yarn, etc.
5. `rails new —webpack —skip-hotwire` is the same as above, minus the default configuration of Hotwire.

In line with the upcoming Webpacker 6, we’ve dropped the —webpack=framework preconfiguration. —webpack is now a boolean flag.

Alpha preview of what this feels like: https://www.youtube.com/watch?v=PtxZvFnL2i0&t=334s

Commit:
Turbolinks is being replaced with Hotwire

Commit:
Make --webpack opt-in

Commit:
Don't use specific webpacker installers any more in preparation for next Webpacker

ghiculescu commented on railties/lib/rails/app_updater.rb:
```diff
@@ -21,14 +21,14 @@ def app_generator
           options[:skip_sprockets]      = !defined?(Sprockets::Railtie)
           options[:skip_bootsnap]       = !defined?(Bootsnap)
           options[:skip_spring]         = !defined?(Spring)
+          options[:webpack]             = File.exist?(Rails.root.join("bin", "yarn"))
```
```suggestion
          options[:webpack]             = File.exist?(Rails.root.join("config", "webpacker.yml"))
```

not sure how common this is but if someone's using npm instead of yarn this wouldn't cover them.

ghiculescu replied:
ps. very excited for this PR. 👋  webpack

Commit:
Update railties/lib/rails/app_updater.rb

Co-authored-by: Alex Ghiculescu <alex@tanda.co>

Commit:
Merge branch 'main' into replace-webpack-with-importmapped-hotwire-as-default-js

Commit:
Merge branch 'main' into replace-webpack-with-importmapped-hotwire-as-default-js

Commit:
Trailing whitespace

Commit:
Convert to Turbo data attribute for tracking

Commit:
Default is no webpack, no hotwire

Commit:
Swap out turbolinks references for hotwire

Commit:
Drop explicit return

Commit:
Merge branch 'main' into replace-webpack-with-importmapped-hotwire-as-default-js

Commit:
Only generate package.json if using webpack

Commit:
Only create package.json in webpack mode

Commit:
Only create app/javascript in webpack mode

Commit:
Generate correct style/js links based on js mode

Commit:
Fix tests from changed output format

Not sure why these are showing up in this PR, though.

Commit:
Rubocopping

Commit:
Stick with webpack for the test app for now

Commit:
Adjust tests

dhh commented:
If anyone would like to help with this effort, I’d appreciate some assistance fixing all the remaining tests ✌️

dfmoreto commented:
It may not be the best "First Issue" to start collaborating on Rails, but I want to give a try. 
Even if I can't reach anything, at least I read some Rails code 🙂

Commit:
Merge branch 'main' into replace-webpack-with-importmapped-hotwire-as-default-js

andrehjr commented on railties/test/application/test_runner_test.rb:
```diff
@@ -502,15 +502,15 @@ def test_output_inline_by_default
       create_test_file :models, "post", pass: false, print: false
 
       output = run_test_command("test/models/post_test.rb")
-      expect = %r{Running:\n\nF\n\nFailure:\nPostTest#test_truth \[[^\]]+test/models/post_test.rb:6\]:\nwups!\n\nrails test test/models/post_test.rb:4\n\n\n\n}
```
Hey @dhh I was checking this out, and it seems related to minitest-reporters gem. Removing the gem and running the test make most tests pass.

What happened is that on this branch, minitest ended up loading minitest-reporters plugin first. Making the following code https://github.com/rails/rails/blob/main/railties/lib/minitest/rails_plugin.rb#L61-L69 not load  `Rails::TestUnitReporter`. 

There are a [few reports](https://github.com/minitest-reporters/minitest-reporters/issues/247) about similar situations. 

One way to fix is to remove that gem and rework the JUnit artifacts on CI without it. minitest-reporters was added on #35698 in order to generate those junit reports only. Would this be a good way forward?

The other option is making `railties/lib/minitest/rails_plugin.rb` work well with minitest-reporters.

andrehjr committed:
Replace minitest-reporters with minitest-ci (#43016)

minitest-reporters is used to create junit xml reports on CI.

But when it loads before rails minitest plugin makes
`Rails::TestUnitReporter` not being added as a reporter.

minitest-ci is now only loaded at ci and does not interferes with
rails minitest plugins. And keeps junit reports workings

Commit:
Too heavy handed to actually run bundle

Just like we don't auto-migrate

Commit:
Pin js frameworks in importmap

Instead of having importmap preconfigure it.

Commit:
Match updated app/javascript path

Commit:
No need for the explaining comment

dhh commented:
Did a walkthrough of where this is going to take Rails: https://www.youtube.com/watch?v=PtxZvFnL2i0

abhaynikam committed:
Fixes test cases for replace webpack with importmapped Hotwire as default js (#42999)

abhaynikam committed:
Fix rubocop issues

abhaynikam committed:
Fix more railities test cases

abhaynikam committed:
Fix plugin generator railties shared test cases

abhaynikam committed:
Fix Action Text install generator asset pipeline spec

abhaynikam commented:
@dhh PR: #43026 should fix all the test cases. 😊

guilleiguaran committed:
Merge pull request #43026 from abhaynikam/fix-feature-importmap-default-test-cases

Fixes test cases for replace webpack with importmapped Hotwire as def…

Commit:
They're modules, not files

Commit:
Let dev use the latest release as well

So we don't have to replace unexisting dev releases with latest release

Commit:
Make Webpack responsible for generating all the JS files it needs

Webpacker 6 has already moved from app/javascript to app/packs.

Commit:
Don't add rails/ujs by default any longer

All the ajax/form functionality has been superseded by Turbo. The rest lives in a weird inbetween land we need to address through other means.

Commit:
Use new importmap location

Commit:
Merge branch 'main' into replace-webpack-with-importmapped-hotwire-as-default-js

Commit:
Switch to using turbo-rails and stimulus-rails directly

The hotwire-rails gem does not offer enough value for its indirection

Commit:
Use latest Webpacker

Commit:
Prevent version resolution requests from getting swallowed

Commit:
Use ESM syntax for imports

Commit:
Move management of yarn, package.json, etc to Webpacker 6

Commit:
Update for Webpacker 6

Commit:
Move bin/setup addition to Webpacker as well

Commit:
Remove dead tests

Commit:
Bump to Webpacker 6.0.0.rc.2

Commit:
No longer relevant given the new default is no webpacker

Commit:
Rely on Webpacker 6

Commit:
No longer relevant

Commit:
No longer relevant

Commit:
Make cable channel generator work for both webpacker and importmap setups

Commit:
Fix tests

Commit:
For tests testing importmap way

Commit:
Use Webpacker 6 dummy

Commit:
RuboCopping

la-ruby commented:
very excited to discover this branch.
[I was getting](https://github.com/la-ruby/create-rails-7-app/runs/3424347166?check_suite_focus=true) a "rails aborted! Webpacker configuration file not found" today.

Commit:
One more bump to fix webpack-dev-server

Commit:
Another bump. Hopefully the last one!

Commit:
Also enough to not want turbo tracking on

Commit:
Fix tests

Commit:
Latest

Commit:
Fix tests

Commit:
Fix more tests

Commit:
Fix tests
